enzyme/Enzyme/MLIR/Implementations/FuncAutoDiffOpInterfaceImpl.cpp

```diff
@@ -29,9 +29,117 @@ namespace {
 #include "Implementations/FuncDerivatives.inc"
 } // namespace
 
+static std::optional<func::FuncOp> getContainingFunction(Operation *orig) {
+  Operation *parent;
+  while (parent = orig->getParentOp()) {
+    if (auto func = dyn_cast<func::FuncOp>(parent)) {
+      return std::optional(func);
+    }
+  }
+
+  return std::nullopt;
+}
+
+class AutoDiffCallRev
+    : public ReverseAutoDiffOpInterface::ExternalModel<AutoDiffCallRev,
+                                                       func::CallOp> {
+public:
+  LogicalResult createReverseModeAdjoint(Operation *orig, OpBuilder &builder,
+                                         MGradientUtilsReverse *gutils,
+                                         SmallVector<Value> caches) const {
+    DerivativeMode mode = DerivativeMode::ReverseModeGradient;
+
+    SymbolTable symbolTable = SymbolTable::getNearestSymbolTable(orig);
+
+    func::CallOp callOp = cast<func::CallOp>(orig);
+
+    auto parent = getContainingFunction(orig);
+    if (parent.has_value() &&
+        callOp.getCallee() == parent.value().getSymName()) {
+      // TODO: Recursion chains
+      orig->emitError() << "could not emit adjoint of recursive call at: "
+                        << *orig << "\n";
+      return failure();
+    }
+
+    Operation *callee = symbolTable.lookup(callOp.getCallee());
+    auto fn = cast<FunctionOpInterface>(callee);
+
+    auto narg = orig->getNumOperands();
+    auto nret = orig->getNumResults();
+
+    std::vector<DIFFE_TYPE> RetActivity(nret, DIFFE_TYPE::OUT_DIFF);
+
+    std::vector<DIFFE_TYPE> ArgActivity;
+    for (auto arg : callOp.getOperands()) {
+      ArgActivity.push_back(gutils->isConstantValue(arg)
+                                ? DIFFE_TYPE::CONSTANT
+                                : DIFFE_TYPE::OUT_DIFF);
```

Comment:
Active or dup can be determined by casting to the AutoDiffTypeInterface and asking if the type is mutable